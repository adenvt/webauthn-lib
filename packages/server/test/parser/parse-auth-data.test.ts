import parseAttestationObject from '../../src/parser/parse-attestation-object'
import parseAuthData from '../../src/parser/parse-auth-data'
import attestationResponse from '../__sample__/android-safetynet.json'
import assertionResponse from '../__sample__/assertion.json'

test('Parsing Attestation AuthData', () => {
  const attsObject = parseAttestationObject(attestationResponse.response.attestationObject)
  const authData   = parseAuthData(attsObject.authData)

  expect(authData).toHaveProperty('rpIdHash')
  expect(authData).toHaveProperty('flags')
  expect(authData).toHaveProperty('signCount')
  expect(authData).toHaveProperty('attestedCredentialData')
  expect(authData).toHaveProperty('extensions')

  expect(authData.flags).toHaveProperty('UP')
  expect(authData.flags).toHaveProperty('UV')
  expect(authData.flags).toHaveProperty('AT')
  expect(authData.flags).toHaveProperty('ED')

  expect(authData.attestedCredentialData).toHaveProperty('aaguid')
  expect(authData.attestedCredentialData).toHaveProperty('credentialIdLength')
  expect(authData.attestedCredentialData).toHaveProperty('credentialId')
  expect(authData.attestedCredentialData).toHaveProperty('credentialPublicKey')

  expect(authData.extensions).toBe(undefined)
  expect(authData).toMatchSnapshot()
})

test('Parsing Assertion AuthData', () => {
  const authData = parseAuthData(Buffer.from(assertionResponse.response.authenticatorData))

  expect(authData).toHaveProperty('rpIdHash')
  expect(authData).toHaveProperty('flags')
  expect(authData).toHaveProperty('signCount')
  expect(authData).toHaveProperty('extensions')

  expect(authData.flags).toHaveProperty('UP')
  expect(authData.flags).toHaveProperty('UV')
  expect(authData.flags).toHaveProperty('AT')
  expect(authData.flags).toHaveProperty('ED')

  expect(authData).toMatchSnapshot()
})

test('Throw error when authData less than 37 bytes', () => {
  expect(() => {
    parseAuthData(Buffer.from('hello world'))
  }).toThrowError('Authenticator Data must be at least 37 bytes long!')
})
