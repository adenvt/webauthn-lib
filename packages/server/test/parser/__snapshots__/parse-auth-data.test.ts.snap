// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsing Assertion AuthData 1`] = `
Object {
  "attestedCredentialData": undefined,
  "extensions": undefined,
  "flags": Object {
    "AT": true,
    "ED": false,
    "UP": true,
    "UV": true,
  },
  "rpIdHash": Object {
    "data": Array [
      83,
      90,
      89,
      78,
      53,
      89,
      103,
      79,
      106,
      71,
      104,
      48,
      78,
      66,
      99,
      80,
      90,
      72,
      90,
      103,
      87,
      52,
      95,
      107,
      114,
      114,
      109,
      105,
      104,
      106,
      76,
      72,
    ],
    "type": "Buffer",
  },
  "signCount": 1450867317,
}
`;

exports[`Parsing Attestation AuthData 1`] = `
Object {
  "attestedCredentialData": Object {
    "aaguid": "00000000000000000000000000000000",
    "credentialId": Object {
      "data": Array [
        1,
        144,
        251,
        134,
        236,
        25,
        87,
        30,
        218,
        91,
        87,
        159,
        69,
        174,
        148,
        171,
        114,
        83,
        64,
        218,
        43,
        143,
        122,
        128,
        247,
        42,
        203,
        32,
        213,
        196,
        114,
        11,
        194,
        65,
        139,
        86,
        137,
        36,
        26,
        213,
        227,
        136,
        86,
        181,
        223,
        10,
        104,
        169,
        232,
        204,
        207,
        240,
        166,
        136,
        76,
        235,
        209,
        46,
        60,
        183,
        77,
        65,
        206,
        239,
        247,
      ],
      "type": "Buffer",
    },
    "credentialIdLength": 65,
    "credentialPublicKey": Object {
      "data": Array [
        165,
        1,
        2,
        3,
        38,
        32,
        1,
        33,
        88,
        32,
        85,
        17,
        100,
        165,
        37,
        234,
        203,
        109,
        27,
        169,
        114,
        240,
        90,
        92,
        14,
        66,
        170,
        154,
        117,
        5,
        174,
        58,
        29,
        148,
        199,
        249,
        218,
        160,
        153,
        227,
        2,
        97,
        34,
        88,
        32,
        137,
        115,
        96,
        179,
        215,
        105,
        86,
        175,
        197,
        186,
        89,
        224,
        110,
        153,
        37,
        98,
        122,
        135,
        5,
        191,
        54,
        8,
        138,
        0,
        122,
        82,
        52,
        191,
        1,
        193,
        255,
        221,
      ],
      "type": "Buffer",
    },
  },
  "extensions": undefined,
  "flags": Object {
    "AT": true,
    "ED": false,
    "UP": true,
    "UV": true,
  },
  "rpIdHash": Object {
    "data": Array [
      149,
      105,
      8,
      143,
      30,
      206,
      227,
      35,
      41,
      84,
      3,
      93,
      189,
      16,
      215,
      202,
      227,
      145,
      48,
      90,
      39,
      81,
      181,
      89,
      187,
      143,
      215,
      203,
      178,
      41,
      189,
      212,
    ],
    "type": "Buffer",
  },
  "signCount": 0,
}
`;
